python/ray/experimental/channel/torch_tensor_type.py

```diff
@@ -1,65 +1,99 @@
+
+
+def _get_default_torch_device() -> "torch.device":
+    from ray.air._internal import torch_utils
+
+    if not ray.get_gpu_ids():
+        import torch
+
+        # torch_utils defaults to returning GPU 0 if no GPU IDs were assigned
+        # by Ray. We instead want the default to be CPU.
+        return torch.device("cpu")
+
+    return torch_utils.get_devices()[0]
 
 
 @PublicAPI(stability="alpha")
 class TorchTensorType(ChannelOutputType):
+    AUTO = "auto"
     NCCL = "nccl"
 
     def __init__(
-        self, shape: Tuple[int], dtype: "torch.dtype", transport: Optional[str] = None
+        self,
+        shape: Union[int, Tuple[int], str] = AUTO,
+        dtype: "torch.dtype" = AUTO,
+        transport: Optional[str] = None,
     ):
+        """
+        A type hint that can be used to annotate DAG nodes that return a
+        torch.Tensor.
+
+        NOTE: Use of this type in the DAG will register a custom serializer for
+        torch.Tensor that moves the tensor to the correct device on the
+        receiver. If you are using ray.cloudpickle to serialize objects and you
+        do not want this behavior, you should deregister the custom serializer
+        using ray.util.serialization.deregister_serializer(torch.Tensor).
+
+        Args:
+            shape: The expected shape of the torch.Tensor. "auto" (default)
+                means that the shape will be dynamically inferred. For tensors
+                passed via host memory (default), the shape is a hint for the
+                maximum size of the tensor. If a DAG node's returned serialized
+                tensor exceeds this size, the task will error. For tensors
+                passed via NCCL, the returned tensor must *match* the given
+                shape; if it does not match, the task will error.
+            dtype: The expected dtype of the torch.Tensor. Similar to the
+                shape, this may be statically or dynamically declared.
+            transport: "auto" (default) means that tensors will be passed via
+                host memory, using numpy as the serialization format. Pass
+                TorchTensorType.NCCL or "nccl" to use NCCL instead, avoiding
+                the host memory copy.
+        """
+        super().__init__()
+
+        if isinstance(shape, str):
+            shape = shape.lower()
+        if isinstance(dtype, str):
+            dtype = dtype.lower()
         self.shape = shape
         self.dtype = dtype
         self.transport = transport
+        self.nccl_group_id: Optional[str] = None
+
+    def register_custom_serializer(self) -> None:
+        super().register_custom_serializer()
 
-    @staticmethod
-    def register_custom_serializer(outer: Any) -> None:
-        # Helper method to run on the DAG driver and actors to register custom
-        # serializers.
         import torch
 
-        from ray.air._internal import torch_utils
+        default_device = _get_default_torch_device()
```

Comment:
probably need to make it flexible or only allow 1 gpu (maybe add assert? assert get_num_gpus() == 1)

Reply:
Yeah, we should make it possible to override in the future.

It will only return 1 GPU right now.